File: 2021/day-03/day03-1.js
const fs = require('fs');

let counts = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
let gamma = 0;
let epsilon = 0;
let total = 0;
const datas = fs.readFileSync('input.txt', 'utf-8').split(/\n/).map(line => {
    ++total;
    for (let index = 0; index < line.length; ++index) {
        counts[index] += parseInt(line[index]);
      }
});

let half = total / 2;
counts.map( count => {
    gamma <<= 1;
    epsilon <<= 1;
    if (count > half) {
        ++gamma;
    } else {
        ++epsilon;
    }
});
console.log(gamma * epsilon);

File: 2021/day-03/day03-1.cpp
#include <iostream>
#include <fstream>
#include <vector>

int main(int argc, char** argv)
{
    std::ifstream file("input.txt");
    std::vector<int> counts {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    std::string line;
    auto total = 0;
    while (file >> line)
    {
        ++total;
        for (auto i = 0; i < line.length(); ++i)
        {
            auto ch = line[i];
            if (ch == '1') {
                ++counts[i];
            }
        }
    }

    auto gamma = 0;
    auto epsilon = 0;
    total >>= 1;
    for (auto count : counts)
    {
        gamma <<= 1;
        epsilon <<= 1;
        if (count >= total)
        {
            gamma += 1;
        }
        else
        {
            epsilon += 1;
        }
    }

    std::cout << gamma * epsilon;

    return 0;
}
